File: ios/Sources/SafeSDKPlugin/SafeSDKPlugin.swift
import Foundation
import Capacitor
import AntifraudSDK
import CoreLocation

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(SafeSDKPlugin)
public class SafeSDKPlugin: CAPPlugin, CAPBridgedPlugin {
    private var locationManager = CLLocationManager()
    public let identifier = "SafeSDKPlugin"
    public let jsName = "SafeSDK"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "initializeWithAccessToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "subscribe", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "unsubscribe", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "check", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setUserId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setAccessToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "askForLocationPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "enableLogging", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = SafeSDK()

    @objc func initialize(_ call: CAPPluginCall) {
        guard let apiBaseUrl = SecretsConfigHelper.getApiBaseUrl(), let apiKey = SecretsConfigHelper.getApiKey(), let authTokenUrl = SecretsConfigHelper.getAuthTokenUrl(), let clientId = SecretsConfigHelper.getClientId(), let clientSecret = SecretsConfigHelper.getClientSecret() else {
            call.reject("Missing required initialize parameters")
            return
        }

        let params = call.getObject("params") as? [String: Any]

        DispatchQueue.main.async {
            self.implementation.initialize(
                apiBaseUrl: apiBaseUrl,
                apiKey: apiKey,
                authTokenUrl: authTokenUrl,
                clientId: clientId,
                clientSecret: clientSecret,
                params: params
            )
        }

        call.resolve()
    }

    @objc func initializeWithAccessToken(_ call: CAPPluginCall) {
        guard let apiBaseUrl = SecretsConfigHelper.getApiBaseUrl(), let apiKey = SecretsConfigHelper.getApiKey() else {
            call.reject("Missing required initialize parameters")
            return
        }

        let params = call.getObject("params") as? [String: Any]
        guard let accessToken = call.getString("accessToken"), !accessToken.isEmpty else {
            call.reject("Missing required accessToken parameter")
            return
        }

        DispatchQueue.main.async {
            self.implementation.initialize(
                apiBaseUrl: apiBaseUrl,
                apiKey: apiKey,
                accessToken: accessToken,
                params: params
            )
        }

        call.resolve()
    }

    @objc func subscribe(_ call: CAPPluginCall) {
        implementation.subscribe { result in
            switch result {
            case .success(let action):
                let actionDict: [String: Any] = [
                    "type": self.mapActionType(actionType: action.actionType),
                    "signature": action.signature
                ]
                self.notifyListeners("receivedAction", data: actionDict)

            case .failure(let error):
                let code = self.mapError(error: error)
                let errorDict: [String: Any] = [
                    "code": code,
                    "message": "SafeSDK subscribe failed: \(code)"
                ]
                self.notifyListeners("subscribeError", data: errorDict)
            }
        }

        call.resolve()
    }

    @objc func check(_ call: CAPPluginCall) {
        implementation.check { result in
            switch result {
            case .success(let action):
                let actionDict: [String: Any] = [
                    "type": self.mapActionType(actionType: action.actionType),
                    "signature": action.signature
                ]
                call.resolve(["action": actionDict])
            case .failure(let error):
                let code = self.mapError(error: error)
                let message = "SafeSDK check failed: \(code)"
                call.reject(message, code, nil)
            }
        }
    }

    @objc func unsubscribe(_ call: CAPPluginCall) {
        implementation.unsubscribe()
        call.resolve()
    }

    @objc func setUserId(_ call: CAPPluginCall) {
        guard let userId = call.getString("userId") else {
            call.reject("Missing required userId parameter")
            return
        }
        implementation.setUserId(userId)
        call.resolve()
    }

    @objc func setAccessToken(_ call: CAPPluginCall) {
        guard let accessToken = call.getString("accessToken") else {
            call.reject("Missing required accessToken parameter")
            return
        }
        implementation.setAccessToken(accessToken)
        call.resolve()
    }

    @objc func askForLocationPermissions(_ call: CAPPluginCall) {
        self.locationManager.requestWhenInUseAuthorization()
        call.resolve()
    }

    @objc func enableLogging(_ call: CAPPluginCall) {
        call.unimplemented("Not implemented on iOS. Use the `debugLoggingEnabled` boolean parameter in the `initialize(_)` method to enable debug logging")
    }

    private func mapActionType(actionType: ActionType) -> String {
        switch actionType {
        case .block:
            return "BLOCK"
        case .limitAccess:
            return "LIMIT_ACCESS"
        case .monitor:
            return "MONITOR"
        case .notSafe:
            return "NOT_SAFE"
        case .safe:
            return "SAFE"
        @unknown default:
            return "DEFAULT_UNKNOWN_ACTION"
        }
    }

    private func mapError(error: SDKError) -> String {
        switch error {
        case .networkConnection:
            return "NETWORK_CONNECTION"
        case .noActiveApiKey:
            return "NO_ACTIVE_API_KEY"
        case .noChecksAvailableFailure:
            return "NO_CHECKS_AVAILABLE"
        case .noRecipeFoundFailure:
            return "NO_RECIPE_FOUND"
        case .notInitialized:
            return "NOT_INITIALIZED"
        case .serverError:
            return "SERVER_ERROR"
        case .unknownError:
            return "UNKNOWN_ERROR"
        case .invalidToken:
            return "INVALID_TOKEN"
        case .authError:
            return "AUTH_ERROR"
        case .alreadyInitialized:
            return "ALREADY_INITIALIZED"
        @unknown default:
            return "DEFAULT_UNKNOWN_ERROR"
        }
    }
}

File: android/src/main/java/com/bespot/gatekeepersdkcapacitorplugin/SafeSDKPlugin.kt
package com.bespot.gatekeepersdkcapacitorplugin

import android.Manifest
import com.bespot.shared.core.Failure
import com.getcapacitor.JSObject
import com.getcapacitor.PermissionState
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.getcapacitor.annotation.PermissionCallback


@CapacitorPlugin(
    name = "SafeSDK",
    permissions = [
        Permission(
            alias = "location",
            strings = [
                Manifest.permission.ACCESS_FINE_LOCATION,
                Manifest.permission.ACCESS_COARSE_LOCATION
            ]
        ),
        Permission(
            alias = "storage",
            strings = [
                Manifest.permission.READ_EXTERNAL_STORAGE,
                Manifest.permission.WRITE_EXTERNAL_STORAGE
            ]
        ),
        Permission(
            alias = "media-audio",
            strings = [
                Manifest.permission.READ_MEDIA_AUDIO
            ]
        )
    ]
)
class SafeSDKPlugin : Plugin() {
    private val implementation = SafeSDK()

    @PluginMethod
    fun echo(call: PluginCall) {
        val value = call.getString("value")
        val ret = JSObject()
        ret.put("value", implementation.echo(value))
        call.resolve(ret)
    }

    @PluginMethod
    fun initialize(call: PluginCall) {
        call.unimplemented("Not implemented on Android. Initialization is done at application launch.")
    }

    @PluginMethod
    fun initializeWithAccessToken(call: PluginCall) {
        val accessToken = call.getString("accessToken")
        val apiBaseUrl = resString("antifraud_sdk_api_url")
        val apiKey = resString("antifraud_sdk_key")
        if (apiBaseUrl.isNullOrEmpty() || apiKey.isNullOrEmpty()) {
            return call.reject("Missing required initialize parameters")
        }
        if (accessToken.isNullOrEmpty()) {
            return call.reject("Missing required accessToken parameter")
        }
        implementation.initializeWithAccessToken(apiBaseUrl, apiKey, accessToken)
        call.resolve()
    }

    @PluginMethod
    fun check(call: PluginCall) {
        implementation.check() { result ->
            result.onSuccess { checkResult ->
                val response = JSObject().apply {
                    put("action", JSObject().apply {
                        put("type", checkResult.action.toString())
                        put("signature", checkResult.ticket)
                    })
                }
                call.resolve(response)
            }
            result.onFailure { error ->
                val (code, message) = mapFailureToCodeAndMessage(error as Failure)

                val response = JSObject().apply {
                    put("code", code)
                    put("message", message)
                }
                call.reject(message, code, response)
            }
        }
    }

    @PluginMethod
    fun setUserId(call: PluginCall) {
        val value = call.getString("userId") ?: return call.reject("Missing required 'userId' parameter")
        implementation.setUserId(value)
        call.resolve()
    }

    @PluginMethod
    fun setAccessToken(call: PluginCall) {
        val value = call.getString("accessToken") ?: return call.reject("Missing required 'accessToken' parameter")
        implementation.setAccessToken(value)
        call.resolve()
    }

    @PluginMethod
    fun enableLogging(call: PluginCall) {
        val value = call.getBoolean("debugLoggingEnabled") ?: return call.reject("Missing required boolean 'debugLoggingEnabled'")
        implementation.enableLogging(value)
        call.resolve()
    }

    @PluginMethod
    fun askForLocationPermissions(call: PluginCall) {
        if (getPermissionState("location") != PermissionState.GRANTED) {
            requestPermissionForAlias("location", call, "locationPermsCallback")
        } else {
            call.resolve()
        }
    }

    @PermissionCallback
    private fun locationPermsCallback(call: PluginCall) {
        if (getPermissionState("location") == PermissionState.GRANTED) {
            call.resolve()
        } else {
            call.reject("Location permission denied")
        }
    }

    @PluginMethod
    fun askForStoragePermissions(call: PluginCall) {
        if (getPermissionState("storage") != PermissionState.GRANTED) {
            requestPermissionForAlias("storage", call, "storagePermsCallback")
        } else {
            call.resolve()
        }
    }

    @PermissionCallback
    private fun storagePermsCallback(call: PluginCall) {
        if (getPermissionState("storage") == PermissionState.GRANTED) {
            call.resolve()
        } else {
            call.reject("Storage permission denied")
        }
    }

    @PluginMethod
    fun askForMediaAudioPermissions(call: PluginCall) {
        if (getPermissionState("media-audio") != PermissionState.GRANTED) {
            requestPermissionForAlias("media-audio", call, "mediaAudioPermsCallback")
        } else {
            call.resolve()
        }
    }

    @PermissionCallback
    private fun mediaAudioPermsCallback(call: PluginCall) {
        if (getPermissionState("media-audio") == PermissionState.GRANTED) {
            call.resolve()
        } else {
            call.reject("Media-audio permission denied")
        }
    }

    private fun mapFailureToCodeAndMessage(error: Failure): Pair<String, String> {
        return when (error) {
            is Failure.NetworkConnection -> "NETWORK_CONNECTION" to "Connection Error"
            is Failure.NoActiveApiKey -> "NO_ACTIVE_API_KEY" to "The API Key is either disabled or wrong"
            is Failure.NoChecksAvailableFailure -> "NO_CHECKS_AVAILABLE" to "The server did not find available checks"
            is Failure.NoRecipeFoundFailure -> "NO_RECIPE_FOUND" to "The application does not have a valid recipe"
            is Failure.NotInitialized -> "NOT_INITIALIZED" to "The SDK is not initialized"
            is Failure.ServerError -> "SERVER_ERROR" to "Remote Server Error"
            is Failure.UnknownError -> "UNKNOWN_ERROR" to "Unknown Error (see support section)"
            is Failure.InvalidToken -> "INVALID_TOKEN" to "The access token is invalid or expired"
            is Failure.AuthError -> "AUTH_ERROR" to "Authentication failed"
            is Failure.AlreadyInitialized -> "ALREADY_INITIALIZED" to "The SDK is already initialized"
        }
    }

    // Resolves a host-app string resource by name (e.g. antifraud_sdk_api_url). These are
    // declared by the app.
    private fun resString(name: String): String? {
        val id = context.resources.getIdentifier(name, "string", context.packageName)
        return if (id != 0) context.getString(id) else null
    }
}
